primitives/payment-handler/src/fee_handler.rs

```diff
@@ -220,56 +259,86 @@ where
 		Ok(final_fee)
 	}
 
-	fn deposit_into_treasury(key: &W::AccountId, credit: W::Credit) -> Result<(), DispatchError> {
-		if let Err(_credit) = W::Assets::resolve(key, credit) {
+	fn deposit_into_treasury(
+		key: &WPA::AccountId,
+		credit: fungibles::Credit<AccountId, PaymentAssets>,
+	) -> Result<(), DispatchError> {
+		if let Err(_credit) = WPA::Assets::resolve(key, credit) {
 			// We decide to continue here, because the error has nothing to do with the
 			// account sending the transaction. It would be a bad user experience if
 			// the transaction fails because we can't allocate the fees to the recipient.
 			log::error!(
-				"Could deposit to treasury, it probably doesn't exist, burning the credit..."
+				"Could not deposit to treasury, it probably doesn't exist, burning the credit..."
 			);
 		}
 		Ok(())
 	}
 }
 
-pub struct NativeGameFeeHandler<AccountId, Balances, WithdrawAsset, Affiliate, Tournament> {
-	_phantom: PhantomData<(AccountId, Balances, WithdrawAsset, Affiliate, Tournament)>,
+pub struct NativeGameFeeHandler<
+	AccountId,
+	PaymentAsset,
+	WithdrawPaymentAsset,
+	Affiliate,
+	AffiliateMaxDistribution,
+	Tournament,
+> {
+	_phantom: PhantomData<(
+		AccountId,
+		PaymentAsset,
+		WithdrawPaymentAsset,
+		Affiliate,
+		AffiliateMaxDistribution,
+		Tournament,
+	)>,
 }
 
-impl<AccountId, Balances, W, Affiliate, Tournament> FeeHandler
-	for NativeGameFeeHandler<AccountId, Balances, W, Affiliate, Tournament>
-where
+impl<AccountId, PaymentAsset, WPA, Affiliate, AffiliateMaxDistribution, Tournament> FeeHandler
+	for NativeGameFeeHandler<
+		AccountId,
+		PaymentAsset,
+		WPA,
+		Affiliate,
+		AffiliateMaxDistribution,
+		Tournament,
+	> where
 	// This is satisfied by the `pallet-balances`.
-	Balances: fungible::Inspect<AccountId, Balance = W::Balance> + fungible::Balanced<AccountId>,
+	PaymentAsset:
+		fungible::Inspect<AccountId, Balance = WPA::Balance> + fungible::Balanced<AccountId>,
+	WPA: WithdrawCredit<AccountId = AccountId, Assets = PaymentAsset>,
+	WPA::Credit: IntoFungibleCredit<AccountId, PaymentAsset>,
 
-	W: WithdrawCredit<
+	Affiliate: DistributeFee<
 		AccountId = AccountId,
-		Assets = Balances,
-		AssetId = (),
-		Credit = fungible::Credit<AccountId, Balances>,
+		Balance = WPA::Balance,
+		FeeDistribution = AffiliateFeeDistribution<
+			AccountId,
+			WPA::Balance,
+			AffiliateMaxDistribution,
+		>,
+	>,
+	Tournament: DistributeFee<
+		AccountId = AccountId,
+		Balance = WPA::Balance,
+		FeeDistribution = TournamentFeeDistribution<AccountId, WPA::Balance>,
 	>,
-
-	Affiliate: DistributeFee<AccountId = W::AccountId, Balance = W::Balance>,
-	Tournament:
-		DistributeFee<AccountId = AccountId, Balance = W::Balance, MaxDistributions = ConstU32<1>>,
 {
 	type AccountId = AccountId;
-	type AssetId = ();
-	type Balance = W::Balance;
+	type Payment = WPA::AssetId;
```

Comment:
I would like to add a comment here, that this can be `()` if no vouchers are intended.